packages/validation/src/validators/metadata/validator.ts

```diff
@@ -0,0 +1,174 @@
+        },
+        help: {
+          text: `Add required ${ruleTypeName}. You can hover the key for documentation.`
+        },
+        defaultConfiguration: {
+          parameters: {name: ruleNormalizedName}
+        },
+      });
+    });
+
+    this.setRules(this._rules);
+  }
+
+  async doValidate(resources: Resource[], incremental?: Incremental): Promise<ValidationResult[]> {
+    const results: ValidationResult[] = [];
+    const dirtyResources = incremental ? resources.filter(r => incremental.resourceIds.includes(r.id)) : resources;
+
+    for (const resource of dirtyResources) {
+      for (const rule of this.rules) {
+        const resourceErrors = this.validateResource(resource, rule);
+        results.push(...resourceErrors);
+      }
+    }
+
+    return results;
+  }
+
+  private isLabelRule(ruleName: string) {
+    return ruleName.endsWith('-labels') || ruleName.endsWith('-label');
+  }
+
+  private validateResource(resource: Resource, rule: RuleMetadataWithConfig) {
+    let invalidKeys: string[] = [];
+    let expectedValues: string[] = [];
+    if (rule.id === 'MTD-recommended-labels') {
+      invalidKeys = this.validateRecommendedLabels(resource, rule);
+    } else if (rule.id === 'MTD-custom-labels') {
+      invalidKeys = this.validateCustomLabels(resource, rule);
+    } else if (rule.id === 'MTD-custom-annotations') {
+      invalidKeys = this.validateCustomAnnotations(resource, rule);
+    } else if (rule.name.endsWith('-label') || rule.name.endsWith('-annotation')) {
+      invalidKeys = this.validateDynamicRule(resource, rule);
+      expectedValues = rule.configuration?.parameters ?? [];
+    }
+
+    if (!invalidKeys.length) {
+      return [];
+    }
+
+    const isLabelRule = this.isLabelRule(rule.name);
+    const expectedValuesText = expectedValues.length ? `, expected values: ${expectedValues.join(', ')}` : '';
+
+    return invalidKeys.map(key => this.adaptToValidationResult(
+        resource,
+        ['metadata', isLabelRule ? 'labels' : 'annotations'],
+        rule.id,
+        `Missing valid '${key}' ${isLabelRule ? 'label' : 'annotation'} in '${resource.kind}'${expectedValuesText}.`
+      )
+    ).filter(isDefined)
+  }
+
+  private validateRecommendedLabels (resource: Resource, rule: RuleMetadataWithConfig) {
+    return this.validateMap(
+      resource.content?.metadata?.labels ?? [],
+      rule.defaultConfiguration?.parameters ?? [],
+      []
+    );
+  }
+
+  private validateCustomLabels(resource: Resource, rule: RuleMetadataWithConfig) {
+    return this.validateMap(
+      resource.content?.metadata?.labels ?? [],
+      rule.configuration?.parameters ?? [],
+      []
+    );
+  }
+
+  private validateCustomAnnotations(resource: Resource, rule: RuleMetadataWithConfig) {
+    return this.validateMap(
+      resource.content?.metadata?.annotations ?? [],
```

Comment:
`?? {}`? I'd even add the defaults to the `validateMap`. 

```
private validateMap(actualMap: {[key: string]: any} = {}, expectedKeys: string[] = [], expectedValues: string[] = []): string[] {
```

Reply:
True, that was a bit ugly with all those `??...`. See https://github.com/kubeshop/monokle-core/pull/387/commits/930d827289555cd92b16e5d67f60f7eb2d05f615.